### DISCUSSION TITLE
DS18B20 emulator for Tasmota

### DISCUSSION THREAD

**Discussion by roesl** [arendst/Tasmota#17133] [Ideas]
I need an emulator for Ds18b20 for connecting Tasmota module port to a heating control.
The measured temperature should be sent from an external module or home automation with mqtt.
I it possible to integrat this function in Tasmota.
There is an arduino library already available.

**Comment by Jason2866** [comment]
Afaik there is no such emulator. Dont wait for, since it is a niche use case, i do not believe someone is writting. It is already possible to subscribe to a mqtt topic and use the payload for further use in Tasmota.
https://tasmota.github.io/docs/MQTT/#subscribeunsubscribe

**Reply by roesl** [reply]
There is https://forum.arduino.cc/uploads/short-url/buhmy02gDWz9lsjlOMHYKoSJ2Ox.ino

and
https://github.com/orgua/OneWireHub

available.

**Comment by Jason2866** [comment]
Okay, there is emulating software. But there is actually only one known person wanting this integrated in Tasmota. For getting Temp data measured (or simulated) from "somewhere" in Tasmota there is already the mqtt subscribe way. So no need for low level emulating an ds18b20.
To say it clear none of the known contributors will write an integration for.

> sent from an external module or home automation with mqtt

Is possible with a standard Tasmota version.

**Reply by roesl** [reply]
The data which is send to tasmota via mqtt should be emulated to a esp output which is connected to the ds18b20 input of a heating control for outside temperature.
This is the only way to use outside temperatur control of a floor heating without using a cabled sensor.
Today there are used heating stations in every appartment of buildings with many appartments and there is no way to use a sensor with cable.

**Reply by roesl** [reply]
There are simple controller for 3 way valves with ds18b20 sensors available

**Reply by Jason2866** [reply]
Edit: Ahh now i understand what you mean. You want to feed an ds18b20 input (from an existing dumb controller) with an ESP GPIO output which is simulating an ds18b20.

**Comment by Jason2866** [comment]
An way to implement could be with `Tasmota Client`. The ds18b20 simulator runs on arduino nano which is connected via `Tasmota Client` to an esp running Tasmota. So it is possible to simulate the ds18b20 without any problematic timing problems and code changes in Tasmota.
See examples https://github.com/arendst/Tasmota/tree/development/tools/fw_TasmotaClient_arduino/TasmotaClient/examples

**Comment by roesl** [comment]
I have a battery powered shelly uni with 18b20 which sends temperature with mqtt.
So how to receive the temperature with arduino?
When the gateway is in integrated in tasmota this would be easier.
The tasmota module is also used for bluetooth sensors and switching the valves of the floor heating.

**Reply by Jason2866** [reply]
You have to write the integration for Tasmota ;-)  
If you use Tasmota Client. Tasmota subscribes the mqtt temp message and is communicating to the nano where the ds18b20 simulater is running. For this variant the sketch needs "only" adopted for communicating to Tasmota via Tasmota Client protocol. Less work and no timing / interrupts etc. issues since the simulator can use 100% the nano.
